pennylane/plugins/default_qubit.py

```diff
@@ -388,78 +376,44 @@ def reset(self):
-        before = 2**wires
-        after = 2**(self.num_wires-wires-1)
-        U = np.kron(np.kron(np.eye(before), U), np.eye(after))
-        return U
-
-    def expand_two(self, U, wires):
-        r"""Expand a two-qubit operator into a full system operator.
-
-        Args:
-          U (array): :math:`4\times 4` matrix
-          wires (Sequence[int]): two target subsystems (order matters!)
+        if self.num_wires == 1:
+            # total number of wires is 1, simply return the matrix
+            return U
 
-        Returns:
-          array: :math:`2^n\times 2^n` matrix
-        """
-        if U.shape != (4, 4):
-            raise ValueError('4x4 matrix required.')
-        if len(wires) != 2:
-            raise ValueError('Two target subsystems required.')
+        N = self.num_wires
         wires = np.asarray(wires)
-        if np.any(wires < 0) or np.any(wires >= self.num_wires) or wires[0] == wires[1]:
+
+        if np.any(wires < 0) or np.any(wires >= N) or len(set(wires)) != len(wires):
             raise ValueError('Bad target subsystems.')
 
-        a = np.min(wires)
-        b = np.max(wires)
-        n_between = b-a-1  # number of qubits between a and b
-        # dimensions of the untouched subsystems
-        before = 2**a
-        after = 2**(self.num_wires-b-1)
-        between = 2**n_between
-
-        U = np.kron(U, np.eye(between))
-        # how U should be reordered
-        if wires[0] < wires[1]:
-            p = [0, 2, 1]
-        else:
-            p = [1, 2, 0]
-        dim = [2, 2, between]
-        p = np.array(p)
-        perm = np.r_[p, p+3]
-        # reshape U into another array which has one index per subsystem, permute dimensions, back into original-shape array
-        temp = np.prod(dim)
-        U = U.reshape(dim * 2).transpose(perm).reshape([temp, temp])
-        U = np.kron(np.kron(np.eye(before), U), np.eye(after))
-        return U
-
-    def expand_multi(self, U, wires):
-        r"""Expand a multi-qubit operator into a full system operator.
+        # generate N qubit basis states via the cartesian product
+        tuples = np.stack(np.meshgrid(*[[0, 1] for _ in range(N)]), -1).reshape(-1, N)
+        tuples[:, [0, 1]] = tuples[:, [1, 0]]
 
-        Args:
-          U (array): :math:`2^n \times 2^n` matrix where n = wires.
-          wires (Sequence[int]): Target subsystems (order matters!)
+        # wires not acted on by the operator
+        inactive_wires = list(set(range(N))-set(wires))
 
-        Returns:
-          array: :math:`2^N\times 2^N` matrix. The full system operator.
-        """
-        raise NotImplementedError
+        # expand U to act on the entire system
+        U = np.kron(U, np.identity(2**len(inactive_wires)))
+
+        # move active wires to beginning of the list of wires
+        rearanged_wires = np.array(list(wires)+inactive_wires)
+
+        # convert to computational basis
+        perm = np.sum(2**np.arange(N-1, -1, -1)*tuples[:, rearanged_wires], axis=1)
```

Comment:
Hard to parse what this is doing. I feel like numpy index (un)raveling might make it more clear what is happening?

Reply:
Essentially, I am just converting the basis state bit string into a decimal number. I.e., `[0, 1, 0, 1, 1] = 2^3+2^1+2^0 = 11`.

Reply:
perfect use-case for ravel/unravel

Reply:
I'm not sure I fully follow how it would work with ravel/unravel?

Reply:
I did a Google search, and found an alternative implementation that is slightly faster:
```python
In [47]: %timeit np.sum(2**np.arange(N-1, -1, -1)*tuples[:, rearanged_wires], axis=1)
12.6 µs ± 280 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [48]: %timeit np.right_shift(np.packbits(tuples[:, rearanged_wires], -1), tuples[:, rearanged_wires].size).squeeze()//2**N
7.49 µs ± 133 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

But I feel that the speeds are so similar, that it isn't much point changing.

Reply:
Sorry, it's `ravel_multi_index` that you want
```np.ravel_multi_index([0,1,0,1,1],[2]*5)```

Reply:
Oh that is awesome!

Reply:
```python
In [61]: %timeit np.ravel_multi_index(tuples[:, rearanged_wires].T, [2]*4)
3.46 µs ± 59 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
Also the fastest implementation!

Reply:
yeah the numpy docs for this function do it a disservice. Even I, who knew what I was looking for, still had to scratch my head a bit to implement it